src/eytzinger_benchmark.zig

```diff
@@ -0,0 +1,319 @@
+            layout.values_count,
+            "E",
+            "B",
+            result.wall_time,
+            result.utime,
+            result.cpu_cycles,
+            result.instructions,
+            result.cache_references,
+            result.cache_misses,
+            result.branch_misses,
+        });
+    }
+
+    {
+        var benchmark = try Benchmark.begin();
+        var i: usize = 0;
+        var v: usize = 0;
+        while (i < layout.searches) : (i += 1) {
+            const target = value_picker[v % value_picker.len];
+            const p = page_picker[i % page_picker.len];
+            const page = pages[p];
+            const hit = page.values[binary_search(Key, Val, Val.key_from_value, Val.key_compare, page.values[0..], target)];
+
+            assert(hit.key == target);
+            if (i % pages.len == 0) v += 1;
+        }
+        const result = try benchmark.end(layout.searches);
+        try stdout.print(body_fmt, .{
+            layout.key_size,
+            layout.value_size,
+            layout.keys_count,
+            layout.values_count,
+            "_",
+            "B",
+            result.wall_time,
+            result.utime,
+            result.cpu_cycles,
+            result.instructions,
+            result.cache_references,
+            result.cache_misses,
+            result.branch_misses,
+        });
+    }
+}
+
+const Layout = struct {
+    blob_size: usize, // bytes allocated for all pages
+    key_size: usize, // bytes per key
+    value_size: usize, // bytes per value
+    keys_count: usize, // keys per page (in the summary)
+    values_count: usize, // values per page
+    searches: usize,
+};
+
+fn Value(comptime layout: Layout) type {
+    return struct {
+        pub const max_key = 1 << (8 * layout.key_size) - 1;
+        pub const Key = math.IntFittingRange(0, max_key);
+        const Self = @This();
+        key: Key,
+        body: [layout.value_size - layout.key_size]u8,
+
+        comptime {
+            assert(@sizeOf(Key) == layout.key_size);
+        }
+
+        fn key_from_value(self: Self) Key {
+            return self.key;
+        }
+
+        fn key_from_key(x: Key) Key {
+            return x;
+        }
+
+        fn key_lt(_: void, a: Key, b: Key) bool {
+            return a < b;
+        }
+
+        fn key_compare(a: Key, b: Key) math.Order {
+            return math.order(a, b);
+        }
+    };
+}
+
+const BenchmarkResult = struct {
+    wall_time: u64, // nanoseconds
+    utime: u64, // nanoseconds
+    cpu_cycles: usize,
+    instructions: usize,
+    cache_references: usize,
+    cache_misses: usize,
+    branch_misses: usize,
+};
+
+const PERF = perf.PERF;
+const perf_event_attr = perf.perf_event_attr;
+const perf_event_open = perf.perf_event_open;
+const perf_counters = [_]PERF.COUNT.HW{
+    PERF.COUNT.HW.CPU_CYCLES,
+    PERF.COUNT.HW.INSTRUCTIONS,
+    PERF.COUNT.HW.CACHE_REFERENCES,
+    PERF.COUNT.HW.CACHE_MISSES,
+    PERF.COUNT.HW.BRANCH_MISSES,
```

Comment:
I think it might be interesting also to run with a 128 MiB blob in total, just to see if perhaps the TLB misses are not dominating the benchmark.

Reply:
No impact.